### PR TITLE
drivers: usb: USB device driver for Kinetis USBFSOTG Controller

### PR DESCRIPTION
**Pull request by jfischer-no**
USB device driver for Kinetis USBFSOTG Controller

I had a driver for Kinetis USBFSOTG controller left over here, so I adapted it for zephyr, it is still WIP/RFC. Tested with FRDM-K64F board. The LEDs/gpio code will be removed, it is only for tracing.

TODO:
- [x] it needs better configuration options for endpoints, dependent on usb class application ~~, currently 64 byte buffers are used by default~~ -> see 259e568.
- [x] ~mpu configuration (as workaround it is simply turned off :fist_oncoming: )~~
- [x] ~clock configuration~~
- [x] testing with KW22D512 (#845)
- [x] better use of both (even/odd) endpoint buffers

### CODE REVIEW

**Comment by jfischer-phytec-iot** [comment]
@fvincenzo @MaureenHelm When the MPU is switched on (default on), the USBFSOTG controller signals the DMA error because it, Bus Master 4,  can not access the SRAM (BD table and the buffer are located there). As I see, I can insert a new region into nxp_mpu_regions.c and/or add BM4 Flags to MPU_REGION_xxxxx, is it correct so far?

**Review by MaureenHelm — COMMENTED** [review]
Thanks @jfischer-phytec-iot. This looks pretty good. I wasn't able to get the cdc sample to work on frdm_k64f though; it got stuck waiting for DTR.

> @fvincenzo @MaureenHelm When the MPU is switched on (default on), the USBFSOTG controller signals the DMA error because it, Bus Master 4, can not access the SRAM (BD table and the buffer are located there). As I see, I can insert a new region into nxp_mpu_regions.c and/or add BM4 Flags to MPU_REGION_xxxxx, is it correct so far? see 678c39b

Let's add the flags to MPU_REGION_xxxx (as you've already done) to be consistent with ethernet.

File: drivers/usb/device/usb_kinetis.c [hunk 1]
```diff
@@ -0,0 +1,929 @@
...
+	if (ep & USB_EP_DIR_IN) {
+		return (u8_t)((dev_data.bd_active >>
+			      (16 + KINETIS_ADDR2IDX(ep))) & 1);
+	}
+	return (u8_t)((dev_data.bd_active >> KINETIS_ADDR2IDX(ep)) & 1);
+}
+
+static inline void kinetis_usb_init_clock(void)
+{
+	/* TODO: refactor or use ext/mcux sdk */
```
Comment (jfischer-phytec-iot) [inline drivers/usb/device/usb_kinetis.c]: Done, I wanted to make the clock source configurable, but the IRC48 did not work for me, so for the first the PLL is the only possible clock source.

File: samples/subsys/usb/cdc_acm/README.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+   usb 9-1: new full-speed USB device number 112 using uhci_hcd
+   usb 9-1: New USB device found, idVendor=8086, idProduct=f8a1
+   usb 9-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
+   usb 9-1: Product: CDC-ACM
+   usb 9-1: Manufacturer: Intel
+   usb 9-1: SerialNumber: 00.01
+   cdc_acm 9-1:1.0: ttyACM1: USB ACM device
+
+The app prints on serial output, used for the console:
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: You don't say how to build (and flash) the sample app.

File: samples/subsys/usb/cdc_acm/README.rst [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+.. code-block:: console
+
+   Wait for DTR
+
+Open a serial port emulator, for example minicom
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: Don't you need to open minicom BEFORE you run the sample app (where does "Wait for DTR" mentioned above appear?)

File: samples/subsys/usb/cdc_acm/README.rst [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+and attach it to detected CDC ACM device:
+
+.. code-block:: console
+
+   minicom --device /dev/ttyACM1
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: default baud rate is ok?

File: samples/subsys/usb/cdc_acm/README.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
+.. _usb_cdc_acm:
+
+USB CDC ACM Sample Application
+####################################
+
+Overview
+********
+
+Sample app for USB CDC ACM driver. The received data is echoed back
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: This sample app demonstrates use of a USB Communication Device Class (CDC) Abstract Control Model (ACM) driver provided by the Zephyr project.  Received data is echoed back to the serial port provided by this driver.

File: samples/subsys/usb/cdc_acm/README.rst [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+Requirements
+************
+
+This project requires an USB device driver,
+which is available for the :ref:`arduino_101` and :ref:`frdm_k64f` boards.
+
+Building and Running
+********************
+
+Plug the board into a host device, for example, a PC running Linux.
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: You don't say how to build (and flash) the sample app.

File: samples/subsys/usb/cdc_acm/README.rst [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+And on ttyACM device, provided by zephyr USB device stack:
+
+.. code-block:: console
+
+   Send characters to the UART device
+   Characters read:
+
+The characters entered in serial port emulator will be echoed back.
+You may need to stop modemmanager, if it is trying to access the
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: You should mention this issue earlier in the steps -- before plugging the board to your host system I suspect.

File: samples/subsys/usb/cdc_acm/README.rst [hunk 8]
```diff
@@ -0,0 +1,63 @@
+.. _usb_cdc_acm:
+
+USB CDC ACM Sample Application
+####################################
+
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: Does this sample app only work on specific boards?  If so, you should list them here.

File: samples/subsys/usb/cdc_acm/README.rst [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+********
+
+Sample app for USB CDC ACM driver. The received data is echoed back
+to the serial port provided by Zephyr CDC ACM driver.
+
+Requirements
+************
+
+This project requires an USB device driver,
+which is available for the :ref:`arduino_101` and :ref:`frdm_k64f` boards.
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: Is this the list of supported boards then?

File: samples/subsys/usb/cdc_acm/README.rst [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+Building and Running
+********************
+
+Plug the board into a host device, for example, a PC running Linux.
+The board will be detected as a CDC ACM device.
```
Comment (dbkinder) [inline samples/subsys/usb/cdc_acm/README.rst]: (as shown by the Linux `dmesg` command):

**Review by dbkinder — DISMISSED** [review]
suggested edits and clarifications

File: samples/subsys/usb/cdc_acm/README.rst (hunk 9 above)
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/cdc_acm/README.rst]: yes, but, I've just seen that the default board for this sample is quark_se_c1000_devboard.

File: samples/subsys/usb/cdc_acm/README.rst (hunk 3 above)
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/cdc_acm/README.rst]: no, the sample waits for DTR, it appears after the minicom opens the device provided by the Zephyr.

**Comment by jfischer-phytec-iot** [comment]
@dbkinder I moved 4a4c6a0 and d8c63e1 to #825

**Comment by jfischer-phytec-iot** [comment]
rebased on master, tested with CDC ACM and DFU, usable but still WIP (leds and blocking while must be removed)

**Comment by galak** [comment]
@jfischer-phytec-iot since still wip, added that to subject, also used github markup for the todo list.  (might need a bit of cleaning up)

**Comment by jfischer-phytec-iot** [comment]
@finikorg confirm, have seen it too, I will investigate

**Comment by finikorg** [comment]
Need to also add this:
```
diff --git a/arch/arm/soc/nxp_kinetis/k6x/soc.h b/arch/arm/soc/nxp_kinetis/k6x/soc.h
index c6fa930c6..a509c0b6c 100644
--- a/arch/arm/soc/nxp_kinetis/k6x/soc.h
+++ b/arch/arm/soc/nxp_kinetis/k6x/soc.h
@@ -35,6 +35,7 @@ extern "C" {
 
 #define IRQ_SPI0 26
 #define IRQ_SPI1 27
+#define IRQ_USB_OTG 53
 #define IRQ_GPIO_PORTA 59
 #define IRQ_GPIO_PORTB 60
 #define IRQ_GPIO_PORTC 61

```

**Comment by jfischer-phytec-iot** [comment]
I agreed to remove the IRQ in the hope that USB DTS support will not take so long, see #4376

**Comment by jfischer-phytec-iot** [comment]
>[usb/usbfsotg] [ERR] usb_dc_ep_write: Endpoint is occupied by the controller

@finikorg it is so far ok, the error appears because previous IN packet was not picked up, so there is no IN token from the USB host as long as host has not configured usb ethernet device.

Unfortunately, I do not see any further write accesses for ep 0x83 and I do not know what the samples really do, with netusb FRDM-K64F has two ethernet interfaces....

**Comment by finikorg** [comment]
@jfischer-phytec-iot So if you disable mcux ethernet on FRDM-K64F then you can ping echo_server? I was trying the same but something was not working, I will try again.

**Comment by jfischer-phytec-iot** [comment]
@finikorg Yes, but then every second outgoing packet will be lost, I suspect IN buffer handling in the driver.

**Comment by jfischer-phytec-iot** [comment]
@finikorg I reworked the driver, see fixup f6adb0a. Tested samples: cdc_acm, mass, webusb and http_server with the script below:

```
#!/bin/bash

IP=$1

for (( n=0; n<1; n++ ))
do
        for (( packetsize=0; packetsize<1600; packetsize=$packetsize+32 ))
        do
                echo "ping payload $packetsize"
                ping -i 0.002 -c 42 -s $packetsize -w 1 $IP
                sleep 1
        done
done

```
packetsize 40 bytes:
rtt min/avg/max/mdev = 0.611/1.216/1.275/0.103 ms

packetsize 1480 bytes:
rtt min/avg/max/mdev = 3.884/3.940/4.154/0.067 ms

**Comment by jfischer-phytec-iot** [comment]
It is waiting for DT support, #4376

**Review by finikorg — APPROVED** [review]
LGTM, tested CDC ACM and ECM functionality on FRDM-K64F

**Comment by jfischer-phytec-iot** [comment]
@nashif ping

**Comment by nashif** [comment]
> @nashif ping

@jfischer-phytec-iot pong.

@MaureenHelm have you looked at this? I assume it is ready to be re-reviewed/tested now.

**Review by MaureenHelm — COMMENTED** [review]
LGTM, but I'd like to get someone from my USB team to also take a look.

I tested CDC on frdm_k64f and it works fine. Do the other samples work too?

**Comment by jfischer-phytec-iot** [comment]
@MaureenHelm 
>I tested CDC on frdm_k64f and it works fine. Do the other samples work too?

Yes, this is my reference driver for usb development (netusb related samples need CONFIG_ETH_MCUX_0 to be disabled).

**Comment by MaureenHelm** [comment]
@jfischer-phytec-iot I'm sorry for the delay, but I got some feedback from our USB team. Here's what they said:

1.  Some SOC configure code is in kinetis_usb_init(void) like:
   /* enable USB voltage regulator */ 
   SIM->SOPT1 |= SIM_SOPT1_USBREGEN_MASK;
   For different SoCs, the register to enable USB regulator might be different. This code only works for a part of Kinetis devices.
 
2. In the Kinetis USB controller, there is limitation that if the device is not ready to receive the setup token sent by the host, the device controller would hang. To avoid such issue, the driver needs to prime a transfer to receive the next setup token as soon as possible once the previous setup transfer is done. From the code we could see, the next setup prime operation would be called by the upper layer of the USB Kinetis driver, it might be not quick enough in some extreme cases.
 
3. The driver is using memcpy operation in the data transfer, which will cause huge throughput performance decrease.
 
4. dev_data.attached is set in function usb_dc_attach(), but it is not clear in usb_dc_detach
 
5. Can’t support multiple KHCI controllers case. Currently there is no multiple KHCI controllers cases in any existed NXP SoC, but from design viewpoint, it is necessary to take it into accounted.
 
6. Any STALL interrupt will cause the STALL status of EP0 be cleared. In case that both EP0 and EP1 should be stalled (sending out STALL handshake), if EP1’s STALL token is sent out and EP1’s STALL interrupt issued, the EP0’s STALL would be cleared before the STALL handshake for EP0 is sent out, then NO STALL handshake for EP0 would be sent out.
 
7. If one EP is in stall status, the transfers on this ep need to be forbidden
 
8. When the Clear_Feature(HALT) is received from the host via EP0, the data toggle needs to be reset. And any unfinished transfer needs to be cancelled. Please take the following case into accounted that the Host sends out a Clear_Feature(HALT) even the corresponding ep is NOT in stall status.

**Comment by jfischer-phytec-iot** [comment]
@nashif @MaureenHelm @finikorg 
Updated, see last commit. I have tested it together with #7435 and the driver passes all tests.

```
./testusb -v 512 -D /dev/bus/usb/005/026                           
unknown speed   /dev/bus/usb/005/026    0
/dev/bus/usb/005/026 test 0,    0.000006 secs
/dev/bus/usb/005/026 test 1,    2.000006 secs
/dev/bus/usb/005/026 test 2,    1.999993 secs
/dev/bus/usb/005/026 test 3,    1.001027 secs
/dev/bus/usb/005/026 test 4,    1.000963 secs
/dev/bus/usb/005/026 test 5,   57.978878 secs
/dev/bus/usb/005/026 test 6,   34.993027 secs
/dev/bus/usb/005/026 test 7,   29.998934 secs
/dev/bus/usb/005/026 test 8,   18.001992 secs
/dev/bus/usb/005/026 test 9,    4.979957 secs
/dev/bus/usb/005/026 test 10,   11.952035 secs
/dev/bus/usb/005/026 test 11,   16.084910 secs
/dev/bus/usb/005/026 test 12,   16.086950 secs
/dev/bus/usb/005/026 test 13,   17.955939 secs
/dev/bus/usb/005/026 test 14,    2.464990 secs
/dev/bus/usb/005/026 test 17,    1.999973 secs
/dev/bus/usb/005/026 test 18,    2.000033 secs
/dev/bus/usb/005/026 test 19,    1.999984 secs
/dev/bus/usb/005/026 test 20,    2.000021 secs
/dev/bus/usb/005/026 test 21,    2.443951 secs
/dev/bus/usb/005/026 test 24,   55.060829 secs
/dev/bus/usb/005/026 test 27,   56.895873 secs
/dev/bus/usb/005/026 test 28,   34.153936 secs
```

Linux kernel output:
```
usbtest 5-1:1.0: TEST 0:  NOP
usbtest 5-1:1.0: TEST 1:  write 1024 bytes 1000 times                                   
usbtest 5-1:1.0: TEST 2:  read 1024 bytes 1000 times                                    
usbtest 5-1:1.0: TEST 3:  write/512 0..1024 bytes 1000 times                            
usbtest 5-1:1.0: TEST 4:  read/512 0..1024 bytes 1000 times                             
usbtest 5-1:1.0: TEST 5:  write 1000 sglists 32 entries of 1024 bytes                   
usbtest 5-1:1.0: TEST 6:  read 1000 sglists 32 entries of 1024 bytes                    
usbtest 5-1:1.0: TEST 7:  write/512 1000 sglists 32 entries 0..1024 bytes               
usbtest 5-1:1.0: TEST 8:  read/512 1000 sglists 32 entries 0..1024 bytes                
usbtest 5-1:1.0: TEST 9:  ch9 (subset) control tests, 1000 times                        
usbtest 5-1:1.0: TEST 10:  queue 32 control calls, 1000 times                           
usbtest 5-1:1.0: TEST 11:  unlink 1000 reads of 1024                                    
usbtest 5-1:1.0: TEST 12:  unlink 1000 writes of 1024                                   
usbtest 5-1:1.0: TEST 13:  set/clear 1000 halts                                         
usbtest 5-1:1.0: TEST 14:  1000 ep0out, 1..1024 vary 512                                
usbtest 5-1:1.0: TEST 17:  write odd addr 1024 bytes 1000 times core map                
usbtest 5-1:1.0: TEST 18:  read odd addr 1024 bytes 1000 times core map                 
usbtest 5-1:1.0: TEST 19:  write odd addr 1024 bytes 1000 times premapped               
usbtest 5-1:1.0: TEST 20:  read odd addr 1024 bytes 1000 times premapped                
usbtest 5-1:1.0: TEST 21:  1000 ep0out odd addr, 1..1024 vary 512                       
usbtest 5-1:1.0: TEST 24:  unlink from 1000 queues of 32 1024-byte writes               
usbtest 5-1:1.0: TEST 27: bulk write 31Mbytes                                           
usbtest 5-1:1.0: TEST 28: bulk read 31Mbytes     
```

**Pull request merged**